extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java

```diff
@@ -36,16 +41,66 @@
 
     private Range range;
 
-    PanacheQueryImpl(EntityManager em, javax.persistence.Query jpaQuery, String query, Object paramsArrayOrMap) {
+    PanacheQueryImpl(EntityManager em, String query, String sort, Object paramsArrayOrMap) {
         this.em = em;
-        this.jpaQuery = jpaQuery;
         this.query = query;
+        this.sort = sort;
         this.paramsArrayOrMap = paramsArrayOrMap;
+
+        if (JpaOperations.isNamedQuery(query)) {
+            String namedQuery = query.substring(1);
+            this.jpaQuery = em.createNamedQuery(namedQuery);
+        } else {
+            this.jpaQuery = em.createQuery(sort != null ? query + sort : query);
+        }
+
+        if (paramsArrayOrMap instanceof Map)
+            JpaOperations.bindParameters(jpaQuery, (Map<String, Object>) paramsArrayOrMap);
+        else
+            JpaOperations.bindParameters(jpaQuery, (Object[]) paramsArrayOrMap);
+
         page = new Page(0, Integer.MAX_VALUE);
     }
 
     // Builder
 
+    @Override
+    public <T> PanacheQuery<T> project(Class<T> type) {
+        if (JpaOperations.isNamedQuery(query)) {
+            throw new PanacheQueryException("Unable to perform a projection on a named query");
+        }
+
+        // We use the first constructor that we found and use the parameter names,
+        // so the projection class must have only one constructor,
+        // and the application must be built with parameter names.
+        // Maybe this should be improved some days ...
+        Constructor<?> constructor = type.getDeclaredConstructors()[0];
+
+        // build select clause with a constructor expression
+        StringBuilder select = new StringBuilder("SELECT new ").append(type.getName()).append(" (");
+        int selectInitialLength = select.length();
+        for (Parameter parameter : constructor.getParameters()) {
+            if (!parameter.isNamePresent()) {
+                throw new PanacheQueryException(
+                        "Your application must be built with parameter names, this should be the default if" +
+                                " using Quarkus artifacts. Check the maven or gradle compiler configuration to include '-parameters'.");
+            }
+
+            if (select.length() > selectInitialLength) {
+                select.append(", ");
+            }
+            select.append(parameter.getName());
+        }
+        select.append(") ");
+
+        this.jpaQuery = em.createQuery(sort != null ? select.toString() + query + sort : select.toString() + query);
+        if (paramsArrayOrMap instanceof Map)
+            JpaOperations.bindParameters(jpaQuery, (Map<String, Object>) paramsArrayOrMap);
+        else
+            JpaOperations.bindParameters(jpaQuery, (Object[]) paramsArrayOrMap);
+        return (PanacheQuery<T>) this;
```

Comment:
Huh, this is a sneaky mutation that changes the type of this query. I'm not sure this is a great idea. Why not return a new query object?

Reply:
PanacheQuery is a statefull object where you can mutate the underlying JPA query object via multiple operations (lock, hint, pagination, ...).
So I stick to the same principle here, projection is just another way to mutate the query before firing it to the database by adding a SELECT clause.

I start thinking that a refactoring should be needed to implements something closer to a builder, each operation on PanacheQuery prior to retrieving the list of result should just update PanacheQuery internal state then when we get the resultList we build the JPA query.

With this, the order of the operation done on the query (see your previous comment) has no importance.

Reply:
And this is what we did for MongoDB with Panache: we create a projection document and change the type of the query. So it's consistent!
https://github.com/quarkusio/quarkus/blob/master/extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/runtime/PanacheQueryImpl.java#L45

Reply:
Well let me show you the problem with changing the runtime type of an object:

```java
PanacheQuery<Order> query = Order.find("name", "Stef");
PanacheQuery<OrderName> otherQuery = query.project(OrderName.class);
Order order = query.firstResult(); // ClassCastException
```

It's fine to mutate the state of an object, but mutating its type is generally not done.

Reply:
I agree this is not convenient but if we create a new `PanacheQuery` for projection it will complexify the implementation and adds maintenance burden in case we want to copy the state of the previous query to the next. Each time we update `PanacheQuery` state management we will need to take care of updating the project method ...

The parameter type of `PanacheQuery` is the return type of the query, this is a component of it's state but I agree this is not very elegant.

I'll take some time to think about this ;)

Reply:
Just make a constructor of `PanacheQueryImpl(PanacheQueryImpl)` which copies the setup and it's all in one place when we add new state. Don't do that in the `project` method.

Reply:
Good idea, I implemented it